go/oasis-test-runner/cmd/root.go

```diff
@@ -73,17 +100,194 @@ func Execute() {
 }
 
+// parseTestParams parses --params.<test_name>.<key1>=<val1>,<val2>... flags combinations, clones provided proto-
+// scenarios, and populates them so that each scenario instance has unique paramater set. Returns mapping test name ->
+// list of scenario instances.
+func parseTestParams(toRun []scenario.Scenario) (map[string][]scenario.Scenario, error) {
+	r := make(map[string][]scenario.Scenario)
+	for _, s := range toRun {
+		zippedParams := make(map[string][]string)
+		for k := range s.Parameters() {
+			userVal := viper.GetStringSlice("params." + s.Name() + "." + k)
+			if userVal == nil {
+				continue
+			}
+			zippedParams[k] = userVal
+		}
+
+		parameterSets := computeParamSets(zippedParams, map[string]string{})
+
+		// For each parameter set combination, clone a scenario and apply user-provided parameter value.
+		for _, ps := range parameterSets {
+			sCloned := s.Clone()
+			for k, userVal := range ps {
+				v := sCloned.Parameters()[k]
+				switch v := v.(type) {
+				case *int:
+					val, err := strconv.ParseInt(userVal, 10, 32)
+					if err != nil {
+						return nil, err
+					}
+					*v = int(val)
+				case *int64:
+					val, err := strconv.ParseInt(userVal, 10, 64)
+					if err != nil {
+						return nil, err
+					}
+					*v = val
+				case *float64:
+					val, err := strconv.ParseFloat(userVal, 64)
+					if err != nil {
+						return nil, err
+					}
+					*v = val
+				case *bool:
+					val, err := strconv.ParseBool(userVal)
+					if err != nil {
+						return nil, err
+					}
+					*v = val
+				case *string:
+					*v = userVal
+				default:
+					return nil, errors.New(fmt.Sprintf("cannot parse parameter. Unknown type %v", v))
+				}
+			}
+			r[s.Name()] = append(r[s.Name()], sCloned)
+		}
+
+		// No parameters provided over CLI, keep a single copy.
+		if len(parameterSets) == 0 {
+			r[s.Name()] = []scenario.Scenario{s}
+		}
+	}
+
+	return r, nil
+}
+
+// computeParamSets recursively combines a map of string slices into all possible key=>value parameter sets.
+func computeParamSets(zp map[string][]string, ps map[string]string) []map[string]string {
+	// Recursion stops when zp is empty. Append ps to result set.
+	if len(zp) == 0 {
+		if len(ps) == 0 {
+			return []map[string]string{}
+		}
+
+		psCloned := map[string]string{}
+		for k, v := range ps {
+			psCloned[k] = v
+		}
+		return []map[string]string{psCloned}
+	}
+
+	rps := []map[string]string{}
+
+	// Take first element from cloned zp and do recursion deterministically.
+	var zpKeys []string
+	for k := range zp {
+		zpKeys = append(zpKeys, k)
+	}
+	sort.Strings(zpKeys)
+
+	zpCloned := map[string][]string{}
+	for _, k := range zpKeys[1:] {
+		zpCloned[k] = zp[k]
+	}
+	for _, v := range zp[zpKeys[0]] {
+		ps[zpKeys[0]] = v
+		rps = append(rps, computeParamSets(zpCloned, ps)...)
+	}
+
+	return rps
+}
+
+// RegisterDefaultScenarios registers all available default scenarios and returns any errors encountered.
+func RegisterDefaultScenarios() (ers []error) {
```

Comment:
Why did you move these here, it looks much less cleaner than before and breaks the separation of the test framework and scenarios.

Reply:
The idea is to recycle the code and use the tests registered in `oasis-test-runner` by `ba` as well.

Reply:
This was fixed by moving `cmp` to `oasis-test-runner` subcommand.